### PR TITLE
kube: add ability to scope to a single namespace

### PR DESCRIPTION
**Pull request by howardjohn**
This follows up on https://github.com/istio/istio/issues/24640, bringing back lost functionality to listen to only a single namespace. This is behind a feature flag, and introduces the ability to watch a single namespace, which is useful for cases where we have multi-tenants running namespace scoped Istiods.

DiscoverySelectors is not sufficient due to the RBAC privileges required.

### CODE REVIEW

**Review by hzxuzhonghu — COMMENTED** [review]
It seems you still need to handle namespace controller. which creates CA in all namespaces.

**Comment by howardjohn** [comment]
> It seems you still need to handle namespace controller. which creates CA in all namespaces.

Good point... the main use case for this I think is to have multiple isolated Istio environment with low privilege so typically they would have that controller disabled. But should handle it

**Comment by imroc** [comment]
IMHO, in the real multi-tenants environment, each tenant usually has multiple namespaces. Currently, when `DiscoverySelectors` are configured, the resources in all namespaces will actually been listed and watched, and filtered on the client side, which still cosume a lot of unnecessary memory and CPU in big multi-tenants clusters.

For example, there are 10 multi-tenant shared clusters, each cluster has 100k pods, that is, a total of 1000k pods, there are 1k tenants, and each tenant has an average of 1k pods (using one or more namespaces across all clusters). Due to the large scale of the cluster, and istiod will list-watch all namespaces of all clusters on startup, which will cause istiod to receive a large amount of network traffic, even if `DiscoverySelectors` will filter out data of other tenants (let GC clean up automatically) , but GC is not real-time, even if a high memory limit is configured for istiod, it is easy to cause OOM(In our real multi-tenant environment, this is a key pain point). And if there are a lot of HPA configuration, auto-scaling occurs almost every minute, istiod still needs to consume a lot of memory continuously.

We did some hack on the code, and it works, istiod mem usage reduced from 10G to 500M on average on startup (varies depending on the number of pods of the tenant), its idea is to only list the namespaces of the tenants in `XXInformer`, and combine the results to return.

I think this idea can be generalized and let istiod natively support it. I noticed that you are refactoring the [client template](https://github.com/istio/istio/blob/e2bff6b5e46993b3eaf84dabc7b78a94434b8289/pkg/config/schema/codegen/collections.go#L353) recently, we can add a similar informer constructor in it, list all concerned ns separately and merge the results and return, watch all concerned ns separately, and combine the results into a new `watch.Interface` and return.

If you think it's a good idea, I can draft a design doc and start implementing it.

**Comment by howardjohn** [comment]
I am personally anti-multi-namespace watches. Kubernetes supports watching 1 namespace or all namespaces. if we want multiple namespaces, we will quickly overwhelm the api-server with watches. If we have a handful (2-5, or so) - not so bad. But if its dynamically changing, adding a removing watches, having 100s of watches *per type* - very suspicious.

**Comment by howardjohn** [comment]
What evidence do we have that 100 namespaces doing this won't cause issues?
Istio watches roughly 20 types, that would be 2k watches per istiod
instances

**Comment by imroc** [comment]
> What evidence do we have that 100 namespaces doing this won't cause issues?

@howardjohn No, I'm not talking about watch, it is list-all.

**Comment by howardjohn** [comment]
Informer is list+watch, isn't it? we never do one without the other

**Comment by howardjohn** [comment]
Oh I see what you mean, we watch all namespaces but list a subset .. I am not sure how that will work with informers

**Comment by howardjohn** [comment]
IMO if this functionality is desired it should be added to k8s core to allow watching multiple namespaces with a single watch

**Comment by imroc** [comment]
> Oh I see what you mean, we watch all namespaces but list a subset

Yes

**Comment by imroc** [comment]
> IMO if this functionality is desired it should be added to k8s core to allow watching multiple namespaces with a single watch

I don't think it's possible. I have seen related issues before, and they think it is appropriate to do it on the client side.

**Comment by hzxuzhonghu** [comment]
Watching multi namespaces separately definitely cause additional load to kube-apiserver, for large scale cluster, this is not acceptable, especially multi-tenant env.  Discovery Selector is not perfect, but I think with strip unused fields, the raw objects watched would not cost much memory.

**Comment by hzxuzhonghu** [comment]
Istio is not for multi tenant design, there are many cluster scoped configurations that could conflict now. Kind of remember webhook configuration is one of those.

**Comment by imroc** [comment]
> Watching multi namespaces separately definitely cause additional load to kube-apiserver, for large scale cluster, this is not acceptable, especially multi-tenant env. Discovery Selector is not perfect, but I think with strip unused fields, the raw objects watched would not cost much memory.

Yep, but now I am not worried about watch, it's the list-all which causes OOM on startup (maybe 20G+ memory with multiple large clusters).

**Comment by imroc** [comment]
> there are many cluster scoped configurations that could conflict now. 

I don't think it's a problem, multi-tenant istio usually have a hidden master cluster, which stores the CRDs per tenant, no conflict.

> Kind of remember webhook configuration is one of those.

And webhook can also have a namespaceSelector

**Comment by imroc** [comment]
Perhaps the best solution for this kind of multi-tenancy of multiple large clusters is to introduce an intermediate component, which performs list-all and watch-all operations on all clusters, and implements the list and watch interfaces of the apiserver, and only returns part of the data to the tenant's istiod which they are interested in, and this is exactly what my team trying to do now.

**Comment by howardjohn** [comment]
> It seems you still need to handle namespace controller. which creates CA in all namespaces.

handled now, thanks

**Review by hzxuzhonghu — APPROVED** [review]

File: pilot/pkg/features/pilot.go [hunk 1]
```diff
@@ -753,6 +753,10 @@ var (
 
 	EnableOptimizedServicePush = env.RegisterBoolVar("ISTIO_ENABLE_OPTIMIZED_SERVICE_PUSH", true,
 		"If enabled, Istiod will not push changes on arbitraty annotation change.").Get()
+
+	InformerWatchNamespace = env.Register("ISTIO_WATCH_NAMESPACE", "",
```
Comment (costinm) [inline pilot/pkg/features/pilot.go:757]: Any chance we can make it a list ( which may have a single element as well ) ? Just for consistency and to not have another change in few months.

File: pkg/kube/client.go [hunk 2]
```diff
@@ -410,37 +411,41 @@ func newClientInternal(clientFactory *clientFactory, revision string) (*client,
 	if err != nil {
 		return nil, err
 	}
-	c.kubeInformer = informers.NewSharedInformerFactory(c.kube, resyncInterval)
+	c.kubeInformer = informers.NewSharedInformerFactoryWithOptions(c.kube, resyncInterval, informers.WithNamespace(features.InformerWatchNamespace))
```
Comment (costinm) [inline pkg/kube/client.go:414]: I guess if it's too complicated to have a list ( multiple informers, etc ) - we can keep it single namespace. Ignore my previous comment.

File: pilot/pkg/features/pilot.go (hunk 1 above)
Comment (howardjohn) [reply pilot/pkg/features/pilot.go:757]: if we add that support it would just be a csv we split I think, no change needed to the user facing variable? lmk if that isn't what you meant

**Comment by costinm** [comment]
On the load/scale issue: I agree that in many K8S installs, having too many watchers is a problem. I also agree that a bigger problem is List operations with HUGE responses, there is paging and all - but it's absurd to send the entire database to a client and filter it on the client. 

I also agree that multi-tenant Istio with v1 API/model is not very viable - injection is a major problem.

However:
- ambient removes injection and most of the high-permission requirements. 
- if K8S can't handle the watches - a caching proxy can be used ( I'm sure they exist - if not we can probably write one in ecosystem, maybe with an extra XDS  adaptation on both sides so it can act as cache for both K8S and XDS and adapt between them). I played a bit with this - I think it's not a big project ( most difficulty is in having some access control).
- IMHO the 'right' pattern for very large scale is on-demand, with very small responses and narrow watches. We can actually do this with ambient. 
- I proposed long ago ( but never got sufficient buy-in) to integrate with CNCF Eventing or similar pubsub adapters, which 
may increase the 'watch' ability to insane scale ( if coupled with an cloud-scale pubsub implementation). This can be bridged in the cache above - and also eliminate a lot of multi-cluster related problems.

TL;DR - Istiod watching multiple namespaces, either by config or on-demand is not bad ( but maybe not in this PR and with more design agreement)

File: pilot/pkg/features/pilot.go (hunk 1 above)
Comment (costinm) [reply pilot/pkg/features/pilot.go:757]: Yes. Maybe an "S" at the end, but we can live with the singular. 

It would still use per-namespace filtered watches - maybe add to the doc that this is intended for either low-privilege OR massive clusters. 

On the interface side - if I remember correctly we have a MeshConfig.configSources list option, one of them is 
a source in k8s://... form. Would it be possible to extend that option with k8s://ns/FOO ? It would also allow expressing 'watch a different cluster', etc. And we already have it.

**Comment by dgn** [comment]
> Perhaps the best solution for this kind of multi-tenancy of multiple large clusters is to introduce an intermediate component, which performs list-all and watch-all operations on all clusters, and implements the list and watch interfaces of the apiserver, and only returns part of the data to the tenant's istiod which they are interested in, and this is exactly what my team trying to do now.

Creating a middleware/proxy component sounds like a good conclusion, I think the same way. But as Costin mentioned, you'll have auth issues. Technically, when the set of namespaces changes, you'd have to make sure to re-authenticate your clients or you could leak data.

To give some context: In OpenShift Service Mesh, we currently support multi-tenant deployments with limited RBAC. Internally, we're using a library called [maistra/xns-informer](https://github.com/maistra/xns-informer) that to istiod "appears" to be a normal InformerFactory and behaves like a global watch, but internally creates separate watches per namespace. It does have performance issues, we regularly see customers with very big meshes (>500 namespaces) creating too much load on the API server.

**Comment by costinm** [comment]
On Thu, Mar 23, 2023 at 9:37 AM Daniel Grimm ***@***.***>
wrote:

> Perhaps the best solution for this kind of multi-tenancy of multiple large
> clusters is to introduce an intermediate component, which performs list-all
> and watch-all operations on all clusters, and implements the list and watch
> interfaces of the apiserver, and only returns part of the data to the
> tenant's istiod which they are interested in, and this is exactly what my
> team trying to do now.
>
> Creating a middleware/proxy component sounds like a good conclusion, I
> think the same way. But as Costin mentioned, you'll have auth issues.
> Technically, when the set of namespaces changes, you'd have to make sure to
> re-authenticate your clients or you could leak data.
>
But not very hard problems - we can still read RBAC or apply policy.
Besides - it is a read-only gateway exposing info that is available anyways
in XDS and Istiod config dumps.

And the proxy is only if the APIserver has scale issues.



> To give some context: In OpenShift Service Mesh, we currently support
> multi-tenant deployments with limited RBAC. Internally, we're using a
> library called maistra/xns-informer
> <https://github.com/maistra/xns-informer> that to istiod "appears" to be
> a normal InformerFactory and behaves like a global watch, but internally
> creates separate watches per namespace. It does have performance issues, we
> regularly see customers with very big meshes (>500 namespaces) creating too
> much load on the API server.
>
That's as expected - and a proxy would help in the big meshes ( don't
forget we also watch multiple clusters ). I'm sure everyone has problems
with
a lot of watches at a certain scale - APIserver is not optimized or really
designed as pubsub.

Not sure if we are just in agreement but proposing different solutions, or
you are advocating for "watch all namespaces and deal with the huge list
response and waste" ?
